src/Compilers/CSharp/Portable/Lowering/LocalRewriter/DelegateCacheRewriter.cs

```diff
@@ -0,0 +1,166 @@
+
+        _factory.Syntax = oldSyntax;
+
+        return rewrittenNode;
+    }
+
+    private DelegateCacheContainer GetOrAddCacheContainer(int localFunctionOrdinal, TypeSymbol delegateType, MethodSymbol targetMethod)
+    {
+        Debug.Assert(_factory.ModuleBuilderOpt is { });
+        Debug.Assert(_factory.CurrentFunction is { });
+
+        var typeCompilationState = _factory.CompilationState;
+        var generation = _factory.ModuleBuilderOpt.CurrentGenerationOrdinal;
+
+        DelegateCacheContainer? container;
+
+        if (CanUseConcreteCacheContainer(delegateType, targetMethod))
+        {
+            container = typeCompilationState.ConcreteDelegateCacheContainer;
+
+            if (container is { })
+            {
+                return container;
+            }
+
+            container = new(typeCompilationState.Type, generation);
+            typeCompilationState.ConcreteDelegateCacheContainer = container;
+        }
+        else
+        {
+            var containersStack = _genericCacheContainers ??= new();
+
+            while (containersStack.Count > 0 && containersStack.Peek().LocalFunctionOrdinal > localFunctionOrdinal)
+            {
+                containersStack.Pop();
+            }
+
+            if (containersStack.Count > 0 && containersStack.Peek().LocalFunctionOrdinal == localFunctionOrdinal)
+            {
+                return containersStack.Peek().CacheContainer;
+            }
+
+            container = new(_factory.CurrentFunction, _topLevelMethodOrdinal, localFunctionOrdinal, generation);
+            containersStack.Push((localFunctionOrdinal, container));
+        }
+
+        _factory.AddNestedType(container);
+
+        return container;
+    }
+
+    private bool CanUseConcreteCacheContainer(TypeSymbol delegateType, MethodSymbol targetMethod)
+    {
+        // Possible places for type parameters that can act as type arguments to construct the delegateType or targetMethod:
+        //   1. containing types
+        //   2. top level method
+        //   3. local functions
+        // Since our containers are created within the same enclosing types, we can ignore type parameters from them.
+
+        Debug.Assert(_factory.CurrentFunction is { });
+
+        // Obviously,
+        if (_topLevelMethod.Arity == 0 && (object)_factory.CurrentFunction == _topLevelMethod)
+        {
+            return true;
+        }
+
+        var methodTypeParameters = PooledHashSet<TypeParameterSymbol>.GetInstance();
+        try
+        {
+            for (Symbol? s = _factory.CurrentFunction; s is MethodSymbol m; s = s.ContainingSymbol)
+            {
+                methodTypeParameters.AddAll(m.TypeParameters);
+            }
+
+            if (methodTypeParameters.Count == 0)
+            {
+                return true;
+            }
+
+            if (targetMethod.MethodKind == MethodKind.LocalFunction)
```

Comment:
>if (targetMethod.MethodKind == MethodKind.LocalFunction) [](http://example.com/codeflow?start=12&length=56)

It looks like we could bail out from the loop above. As soon as we discovered a type parameter.

Reply:
Yes we could, but I think the current shape looks clearer and compact for me. The reason is that, discovering a type parameter doesn't mean we cannot use a concrete(non-generic) cache container, it actually means if we want, we'll have to check a lot more into the local function, instead of now we simply say we cannot. But currently, considered the complexities added, we better off from it and start with simple.

Reply:
>The reason is that, discovering a type parameter doesn't mean we cannot use a concrete(non-generic) cache container, it actually means if we want, we'll have to check a lot more into the local function, instead of now we simply say we cannot. 

Sorry, I am not following. The way I interpret behavior of the code is that, if a target method is a local function and we have any type parameters, we return false. We are not checking anything else. So, what is the point spending time adding things into the HashSet? If later we decide to complicate the rules, we adjust accordingly.

Reply:
Yes, it is. We can ask the same question for the visitor checks later: why spending time adding more into the HashSet if we found targetMethod or delegateType refers to what we currently found. To me the current implementation is more like two phrases: collect information and make decision. That's why I think it's clearer. The loop body is smaller, that's why I think it's compact. We can put everything in the loop and clear the HashSet before the next iteration if you like.

Reply:
>  To me the current implementation is more like two phrases: collect information and make decision. That's why I think it's clearer. The loop body is smaller, that's why I think it's compact. 

I understand, but I would definitely sacrifice compactness to efficiency. Especially given that complexity isn't going to increase much.

> We can put everything in the loop and clear the HashSet before the next iteration if you like.

No, I wouldn't do that. Visiting types once is likely be more efficient than doing it multiple times. If the outcome is the same, I would stick with a single visit at the end. This has no impact on local functions though, those do not need a visit, ever.

Reply:
Good.